GithubRepo/Data/Repository/GithubMainRepository.swift

```diff
@@ -6,47 +6,37 @@
 //
 
 import Foundation
+import RxRelay
+import RxSwift
 
 class GithubMainRepository: GithubRepository {
     private let dataSource: GithubDataSource
-    private(set) var repositories: [Repository]
-    private(set) var loading: Bool
-    private(set) var error: Bool
-    weak var delegate: GithubRepositoryDelegate?
+    private(set) var repositories: BehaviorRelay<[Repository]>
+    private(set) var state: BehaviorRelay<FetchState>
 
     init(dataSource: GithubDataSource = RemoteGithubDataSource()) {
         self.dataSource = dataSource
-        self.repositories = []
-        self.loading = false
-        self.error = false
+        self.repositories = BehaviorRelay<[Repository]>(value: [])
+        self.state = BehaviorRelay<FetchState>(value: .inital)
     }
 
     func handleChangeRepositories(_ repo: [Repository]) {
-        self.repositories = repo
-        delegate?.didChangeRepositories(repositories: repo)
+        self.repositories.accept(repo)
     }
 
-    func handleChangeLoading(_ loading: Bool) {
-        self.loading = loading
-        delegate?.didChangeLoading(loading: loading)
-    }
-
-    func handleChangeError(_ error: Bool) {
-        self.error = error
-        delegate?.didChangeError(error: error)
+    func handleChangeState(_ state: FetchState) {
+        self.state.accept(state)
     }
 
     func fetchRepositories(with query: String) {
-        handleChangeLoading(true)
+        handleChangeState(.loading)
         self.dataSource.getRepositories(with: query) { [weak self] response in
             switch response {
             case .success(let repositoriesResponse):
                 self?.handleChangeRepositories(repositoriesResponse)
-                self?.handleChangeLoading(false)
-                self?.handleChangeError(false)
+                self?.handleChangeState(repositoriesResponse.isEmpty ? .empty: .content)
             case .failure:
-                self?.handleChangeLoading(false)
-                self?.handleChangeError(true)
+                self?.handleChangeState(.error)
             }
```

Comment:
I would like to see you converting all asynchronous code into Rx code. It will be a nice exercise because you gonna need different kinds of subjects/observables/operators... Just for example, if your data layer had the same rx communication, you can easily communicate between components by connecting observables and subjects, like the below improvements comments below that I wrote here for the view layer.

You can do that in the next PR.

Also, take a look in the Rx extensions provided in the RxCommunity GitHub. One good example for you is the rx extensions for Alamofire, called RxAlamofire